### PR TITLE
Update `topology.conf` during `slurmsync`

### PR DESCRIPTION
**Pull request by mr0re1**
Naive implementation that always run `reconfigure` if any changes detected to topology.
No drawbacks since no changes are expected yet.

### CODE REVIEW

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py [hunk 1]
```diff
@@ -522,6 +529,11 @@ def main():
     except Exception:
         log.exception("failed to sync custom scripts")
 
+    try:
+        update_topology(lkp)
+    except Exception:
+        log.exception("failed to sync custom scripts")
```
Comment (nick-stroud) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py:535]: update error message

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/conf.py [hunk 2]
```diff
@@ -482,18 +482,24 @@ def gen_topology(lkp: util.Lookup) -> TopologyBuilder:
...
     conf_file = Path(lkp.cfg.output_dir or slurmdirs.etc) / "cloud_topology.conf"
+    old_hash = util.hash_file(conf_file) if conf_file.exists() else ""
+
     with open(conf_file, "w") as f:
         f.writelines(FILE_PREAMBLE + "\n")
         for line in bldr.render_conf_lines():
             f.write(line)
             f.write("\n")
         f.write("\n")
-    util.chown_slurm(conf_file, mode=0o600)
```
Comment (nick-stroud) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/conf.py:495]: Why is this removed? Seems unrelated.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py [hunk 3]
```diff
@@ -522,6 +529,11 @@ def main():
     except Exception:
         log.exception("failed to sync custom scripts")
 
+    try:
+        update_topology(lkp)
```
Comment (nick-stroud) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py:533]: We are now generating topology on every sync. Description says "no drawbacks since no changes expected".

1. What is cost of generating topology on large cluster, say 5000 nodes. Even though we may not call reconfigure on every sync could this impact sync run time? Should we guard this behind a flag? For example when topology/none is selected then this does not run?
2. Why do we expect it not to change? Probably just a lack of understanding of our current state but seems like whole reason for this feature is we expect it to change.

**Review by nick-stroud — CHANGES_REQUESTED** [review]

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py [hunk 4]
```diff
@@ -488,10 +488,11 @@ def reconfigure_slurm():
...
             log.info("Restarting slurmctld to make changes take effect.")
             try:
+                # TODO: consider removing "restart" since "reconfigure" should restart slurmctld as well
                 run("sudo systemctl restart slurmctld.service", check=False)
-                run(f"{lkp.scontrol} reconfigure", timeout=30)
-            except Exception as e:
-                log.error(e)
+                util.scontrol_reconfigure(lkp)
+            except Exception:
+                log.exception("failed to reconfigure slurmctld")
```
[the remark this answers is not in the document]
Comment (mr0re1) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py:495]: > You are swallowing exception details here. Why? 

No, observe:

```py
import logging
log = logging.getLogger()
def fn(): raise ValueError("gg")

print("=== try log.exception")
try:
  fn()
except Exception:
  log.exception("terrible event")

print("=== try log.error")
try:
  fn()
except Exception as e:
  log.error(e)
```

```sh
=== try log.exception
terrible event
Traceback (most recent call last):
  File "/usr/local/google/home/orlov/tmp/s.py", line 9, in <module>
    fn()
  File "/usr/local/google/home/orlov/tmp/s.py", line 5, in fn
    def fn(): raise ValueError("gg")
              ^^^^^^^^^^^^^^^^^^^^^^
ValueError: gg
=== try log.error
gg
```

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/conf.py (hunk 2 above)
Comment (mr0re1) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/conf.py:495]: `install_topology_conf` does it already. No need to duplicate.

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py (hunk 3 above)
Comment (mr0re1) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py]: > What is cost of generating topology on large cluster, say 5000 nodes.

It's super (cheap < `0.01 sec`) currently, without using `physicalHost`

> Should we guard this behind a flag? For example when topology/none is selected then this does not run?

Good call, added condition to do it IFF `topology/tree` is used.

> Why do we expect it not to change?

Because only possible change (until we use `physicalHost`) is addition/removal of nodes, that in order should be cased by changes in configuration -> full blown reconfigure will happen before `update_topology`

**Review by nick-stroud — APPROVED** [review]
